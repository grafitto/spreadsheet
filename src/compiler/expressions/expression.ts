import { Grid } from '@/types';

export abstract class Expression {
  protected left: Expression | null;
  protected right: Expression | null;

  constructor (left: Expression | null, right: Expression | null) {
    this.left = left;
    this.right = right;
  }
  abstract evaluate(grid: Grid): unknown ;
}