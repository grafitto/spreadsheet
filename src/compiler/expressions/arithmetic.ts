import { Grid } from '../../types';
import { FunctionCallExpression, IdentifierExpression } from '.';
import { Expression } from './expression';

export class ArithmeticExpression extends Expression {
  private operator: string;
  constructor(left: IdentifierExpression, right: ArithmeticExpression | IdentifierExpression | FunctionCallExpression, operator: string) {
    super(left, right);
    this.operator = operator;
  }

  public evaluate(grid: Grid): number {
    switch (this.operator) {
      case '+':
        // @ts-ignore
        return (this.left?.evaluate(grid) as number) + (this.right?.evaluate(grid)) as number;
      case '-':
        // @ts-ignore
        return (this.left?.evaluate(grid) as number) - (this.right?.evaluate(grid)) as number;
      case '*':
        // @ts-ignore
        return (this.left?.evaluate(grid) as number) * (this.right?.evaluate(grid)) as number;
      case '/':
        // @ts-ignore
        return (this.left?.evaluate(grid) as number) / (this.right?.evaluate(grid)) as number;
      default:
        throw Error('Unknown arithmetic operator');

    }
  }
}