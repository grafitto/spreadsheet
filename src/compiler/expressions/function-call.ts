import { Grid } from '@/types';
import { RangeExpression } from '.';
import { Expression } from './expression';

export class FunctionCallExpression extends Expression {
  private name: string;

  constructor(parameters: RangeExpression | Expression, name: string) {
    super(null, parameters);
    this.name = name;
  }

  public evaluate(grid: Grid): number {
    // Should evaluate the right side then apply the function call
    switch(this.name) {
      case 'SUM':
        return this.sum(this.right?.evaluate(grid) as number[]);
      case 'AVG': {
        const paramValues = this.right?.evaluate(grid) as number[];
        return this.sum(paramValues) / paramValues.length;
      }
      default:
        throw Error(`Error evaluating the function call ${this.name}`);
    }
  }

  private sum(items: number[]): number {
    return items.reduce((sum: number, current: number) => sum + current, 0);
  }
}