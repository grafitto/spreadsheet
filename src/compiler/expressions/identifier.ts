import { Grid } from '@/types';
import { Expression } from './expression';
import { alphabeticPositions } from '@/helpers/grid';

export class IdentifierExpression extends Expression {
  private value: string;

  constructor (value: string) {
    super(null, null)
    this.value = value;
  }

  get Value () {
    return this.value;
  }

  /**
   * Fetches the value from the store and returns it
   */
  public evaluate (grid: Grid): unknown {
    const column = alphabeticPositions[this.value[0].toUpperCase()];
    const row = parseInt(this.value[1]) - 1;
    const value = grid[row][column].value;
    if(typeof value === 'number') {
      return value;
    } else {
      throw Error(`Cell ${this.value} has a an unsupported type.`);
    }
  }
}