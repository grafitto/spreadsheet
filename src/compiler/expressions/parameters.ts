import { Grid } from '@/types';
import { IdentifierExpression } from '.';
import { Expression } from './expression';

export class ParametersExpression extends Expression {
  private parameters: (ParametersExpression | IdentifierExpression)[];

  constructor(parameters: (ParametersExpression | IdentifierExpression)[]) {
    super(null, null);
    this.parameters = parameters;
  }

  public evaluate(grid: Grid): unknown {
    const values: number[] = [];
    this.parameters.forEach((parameter: Expression) => {
      values.push(parameter.evaluate(grid) as number);
    });
    return values;
  }
}