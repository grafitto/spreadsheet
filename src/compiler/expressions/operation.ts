import { Grid } from '../../types';
import { FunctionCallExpression, IdentifierExpression } from '.';
import { Token } from '../token';
import { ArithmeticExpression } from './arithmetic';
import { Expression } from './expression';

type ArithmeticOperands = IdentifierExpression | FunctionCallExpression | Token;
export class OperationExpression extends Expression {
  private PRECEDENCE: { [key: string]: number } = {
    '+': 14,
    '-': 14,
    '/': 15,
    '*': 15
  }
  private list: (FunctionCallExpression | IdentifierExpression | Token)[]
  constructor(list: (FunctionCallExpression | IdentifierExpression | Token)[]) {
    super(null, null)
    this.list = list;
  }

  public evaluate(grid: Grid): number {
    const operands: ArithmeticOperands[] = [];
    const operators: Token[] = [];
    if(this.list.length === 1) {
      const identifier: IdentifierExpression = this.list[0] as IdentifierExpression;
      return identifier.evaluate(grid);
    } else {
      for (const item of this.list) {
        if(item instanceof FunctionCallExpression || item instanceof IdentifierExpression) {
          operands.push(item);
        } else {
          // It is a token
          const token = item as Token;
          if(operators.length !== 0) {
            for(let i = 0; i < operators.length; i++) {
              const peek = operators[operators.length - 1];
              if(this.PRECEDENCE[peek.Value] >= this.PRECEDENCE[token.Value]) {
                operands.push(operators.pop() as Token);
              } else {
                break;
              }
            }
          }
          operators.push(token);
        }
      }
      operators.reverse().forEach(operator => {
        operands.push(operator);
      });
    }
    
    return this.constructTree(operands).evaluate(grid);
  }

  public constructTree(operands: ArithmeticOperands[]): ArithmeticExpression {
    const list: (ArithmeticOperands | ArithmeticExpression)[] = [];
    for (const operand of operands) {
      if (operand instanceof Token) {
        const right = list.pop();
        const left = list.pop();
        // @ts-ignore
        list.push(new ArithmeticExpression(left, right, operand.Value as string));
      } else {
        list.push(operand);
      }
    }
    return list[0] as ArithmeticExpression;
  }
  
}