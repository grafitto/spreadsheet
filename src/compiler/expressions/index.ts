import { ProgramExpression } from './program';
import { RangeExpression } from './range';
import { IdentifierExpression } from './identifier';
import { FunctionCallExpression } from './function-call';
import { OperationExpression } from './operation';
import { ParametersExpression } from './parameters';

export {
  RangeExpression,
  ProgramExpression,
  OperationExpression,
  ParametersExpression,
  IdentifierExpression,
  FunctionCallExpression,
}