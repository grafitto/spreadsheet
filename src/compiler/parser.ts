import { Expression } from './expressions/expression';
import { Lexer } from './lexer';
import { TokenType } from './token-type';
import {
  RangeExpression,
  ProgramExpression,
  IdentifierExpression,
  FunctionCallExpression,
} from './expressions';
import { ParametersExpression } from './expressions/parameters';
import { OperationExpression } from './expressions/operation';
import { Token } from './token';
import { ParameterExpression } from '../types';

export class Parser {
  private lexer: Lexer;
  // @ts-ignore
  private program: ProgramExpression;

  constructor (code: string) {
    this.lexer = new Lexer(code);
    this.lexer.generate();
  }

  public parse (): ProgramExpression {
    const token = this.lexer.next();
    while(token.Type !== TokenType.EOF) {
      if(token.Type === TokenType.FUNCTION) {
        this.lexer.previous();
        const functionCall = this.functionCall();
        if (this.lexer.peek().Type ===TokenType.EOF) {
          // Its a single function
          return new ProgramExpression(functionCall);
        } else {
          // Its an expression that begins with a function
          return new ProgramExpression(this.operation(functionCall));

        }
      } else if(token.Type === TokenType.IDENTIFIER) {
        // An operation starting with an identifier
        const identifier = new IdentifierExpression(token.Value as string);
        return new ProgramExpression(this.operation(identifier));
      }
    }
    return this.program;
  }

  private operation (firstOperand: FunctionCallExpression | IdentifierExpression): OperationExpression {
    const list: (FunctionCallExpression | IdentifierExpression | Token)[] = [];
    list.push(firstOperand);
    let token = this.lexer.next();
    while (token.Type !== TokenType.EOF) {
      if (token.IsOperator) {
        list.push(token);
        token = this.lexer.next();
      } else if (token.Type === TokenType.FUNCTION) {
        this.lexer.previous();
        list.push(this.functionCall());
      } else if (token.Type === TokenType.IDENTIFIER) {
        list.push(new IdentifierExpression(token.Value as string));
        token = this.lexer.next();
      }
    }
    return new OperationExpression(list);
  }

  private functionCall (): FunctionCallExpression {
    const name = this.lexer.next();
    this.consume(TokenType.OPEN_BRACKET);
    const parameters = this.functionParameters();
    this.consume(TokenType.CLOSE_BRACKET);
    return new FunctionCallExpression(parameters, name.Value as string);
  }

  private functionParameters (): RangeExpression | ParametersExpression {
    const token = this.lexer.next();
    if (token.Type === TokenType.IDENTIFIER) {
      if (this.lexer.peek().Type === TokenType.FULL_COLLON) {
        // Its a range
        const start = new IdentifierExpression(token.Value as string);
        this.consume(TokenType.FULL_COLLON);
        const end = new IdentifierExpression(this.lexer.next().Value as string)
        return new RangeExpression(start, end);
      } else if (this.lexer.peek().Type === TokenType.COMMA) {
        this.lexer.previous();
        const parameters: ParameterExpression = [];
        while (this.lexer.peek().Type !== TokenType.CLOSE_BRACKET) {
          const innerToken = this.lexer.next();
          if (innerToken.Type === TokenType.IDENTIFIER) {
            parameters.push(new IdentifierExpression(innerToken.Value as string));
          } else if (innerToken.Type === TokenType.FUNCTION) {
            this.lexer.previous();
            parameters.push(this.functionCall())
          }
        }
        return new ParametersExpression(parameters);
      } else {
        throw Error(`Unknown character ${token.RawValue} at position ${this.lexer.Position}`)
      }
    } else if (token.Type === TokenType.FUNCTION) {
      const parameters: ParameterExpression = [];
      this.lexer.previous();
      const functionCall: FunctionCallExpression = this.functionCall() as FunctionCallExpression;
      parameters.push(functionCall);
      if (this.lexer.peek().Type === TokenType.COMMA) {
        while (this.lexer.peek().Type !== TokenType.CLOSE_BRACKET) {
          const innerToken = this.lexer.next();
          if (innerToken.Type === TokenType.COMMA) {
            continue;
          } else if (innerToken.Type === TokenType.IDENTIFIER) {
            parameters.push(new IdentifierExpression(innerToken.Value as string));
          } else if (innerToken.Type === TokenType.FUNCTION) {
            this.lexer.previous();
            parameters.push(this.functionCall())
          }
        }
        return new ParametersExpression(parameters);
      } else {
        throw Error(`Unsupported character ${token.RawValue}`);
      }
    } else {
      throw Error(`Unsupported character ${token.RawValue}`);
    }
  }

  private consume(token: TokenType): void {
    const consumed = this.lexer.next();
    if (consumed.Type !== token) {
      throw Error(`Unexpected token ${consumed.RawValue} at position ${this.lexer.Position}`);
    }
  }
}