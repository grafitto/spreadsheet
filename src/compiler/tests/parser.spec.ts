import { ProgramExpression } from '../expressions';
import { Parser } from '../parser';

describe('Paser', () => {
  let parser: Parser;

  describe('Success', () => {
    describe('Function calls', () => {
      const functions = ['SUM(A1, A2, A3)', 'SUM(A1:A3)', 'SUM(AVG(A1:A3), A5)'];

      functions.forEach((func: string) => {
        it(`Should parse and create the correct syntax tree ${func}`, async () => {
          const code = func;
          parser = new Parser(code);
          const program = parser.parse();
          expect(program).toMatchSnapshot();
          expect(program).toBeInstanceOf(ProgramExpression);
        })
      })
    })
    describe('Arithmetic expressions', () => {
      const expressions = ['A1', 'A1 + A2', 'SUM(A1:A5) - A9'];

      expressions.forEach((expression: string) => {
        it(`Should parse and create the correct syntax tree ${expression}`, async () => {
          const code = expression;
          parser = new Parser(code);
          const program = parser.parse();
          expect(program).toMatchSnapshot();
          expect(program).toBeInstanceOf(ProgramExpression);
        })
      })
    })
  })
})
