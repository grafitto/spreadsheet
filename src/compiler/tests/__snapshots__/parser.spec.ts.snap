// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paser Success Arithmetic expressions Should parse and create the correct syntax tree A1 + A2 1`] = `
ProgramExpression {
  "left": null,
  "right": OperationExpression {
    "left": null,
    "list": Array [
      IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A1",
      },
      Token {
        "rawValue": "+",
        "type": 6,
        "value": "+",
      },
      IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A2",
      },
    ],
    "right": null,
  },
}
`;

exports[`Paser Success Arithmetic expressions Should parse and create the correct syntax tree A1 1`] = `
ProgramExpression {
  "left": null,
  "right": OperationExpression {
    "left": null,
    "list": Array [
      IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A1",
      },
    ],
    "right": null,
  },
}
`;

exports[`Paser Success Arithmetic expressions Should parse and create the correct syntax tree SUM(A1:A5) - A9 1`] = `
ProgramExpression {
  "left": null,
  "right": OperationExpression {
    "left": null,
    "list": Array [
      FunctionCallExpression {
        "left": null,
        "name": "SUM",
        "right": RangeExpression {
          "left": IdentifierExpression {
            "left": null,
            "right": null,
            "value": "A1",
          },
          "right": IdentifierExpression {
            "left": null,
            "right": null,
            "value": "A5",
          },
        },
      },
      Token {
        "rawValue": "-",
        "type": 7,
        "value": "-",
      },
      IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A9",
      },
    ],
    "right": null,
  },
}
`;

exports[`Paser Success Function calls Should parse and create the correct syntax tree SUM(A1, A2, A3) 1`] = `
ProgramExpression {
  "left": null,
  "right": FunctionCallExpression {
    "left": null,
    "name": "SUM",
    "right": ParametersExpression {
      "left": null,
      "parameters": Array [
        IdentifierExpression {
          "left": null,
          "right": null,
          "value": "A1",
        },
        IdentifierExpression {
          "left": null,
          "right": null,
          "value": "A2",
        },
        IdentifierExpression {
          "left": null,
          "right": null,
          "value": "A3",
        },
      ],
      "right": null,
    },
  },
}
`;

exports[`Paser Success Function calls Should parse and create the correct syntax tree SUM(A1:A3) 1`] = `
ProgramExpression {
  "left": null,
  "right": FunctionCallExpression {
    "left": null,
    "name": "SUM",
    "right": RangeExpression {
      "left": IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A1",
      },
      "right": IdentifierExpression {
        "left": null,
        "right": null,
        "value": "A3",
      },
    },
  },
}
`;

exports[`Paser Success Function calls Should parse and create the correct syntax tree SUM(AVG(A1:A3), A5) 1`] = `
ProgramExpression {
  "left": null,
  "right": FunctionCallExpression {
    "left": null,
    "name": "SUM",
    "right": ParametersExpression {
      "left": null,
      "parameters": Array [
        FunctionCallExpression {
          "left": null,
          "name": "AVG",
          "right": RangeExpression {
            "left": IdentifierExpression {
              "left": null,
              "right": null,
              "value": "A1",
            },
            "right": IdentifierExpression {
              "left": null,
              "right": null,
              "value": "A3",
            },
          },
        },
        IdentifierExpression {
          "left": null,
          "right": null,
          "value": "A5",
        },
      ],
      "right": null,
    },
  },
}
`;
